From c892dd539ec765687bee56786a4d06bf1f0a7252 Mon Sep 17 00:00:00 2001
From: Ajith C <quic_ajithc@quicinc.com>
Date: Thu, 22 Aug 2024 15:16:06 +0530
Subject: [PATCH] wifi: ath12k: send valid length when clearing keys

Currently, security keys are not being cleared correctly
when the Access Point switches between security modes.

This issue arises because the firmware expects a valid
key length not only when setting a key but also when
clearing it. However, the driver always sends a key length
of 0 during clear key operations.

Fix this issue by sending the key length along with clear
key commad.

Patch-work: none
Patch-dependency: none

Signed-off-by: Ajith C <quic_ajithc@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/mac.c | 4 +---
 drivers/net/wireless/ath/ath12k/wmi.c | 3 ++-
 2 files changed, 3 insertions(+), 4 deletions(-)

diff --git a/drivers/net/wireless/ath/ath12k/mac.c b/drivers/net/wireless/ath/ath12k/mac.c
index e77813a..1ea86c4 100644
--- a/drivers/net/wireless/ath/ath12k/mac.c
+++ b/drivers/net/wireless/ath/ath12k/mac.c
@@ -6325,9 +6325,7 @@ static int ath12k_install_key(struct ath12k_link_vif *arvif,
 		return 0;
 
 	if (cmd == DISABLE_KEY) {
-		/* TODO: Check if FW expects  value other than NONE for del */
-		/* arg.key_cipher = WMI_CIPHER_NONE; */
-		arg.key_len = 0;
+		arg.key_cipher = WMI_CIPHER_NONE;
 		arg.key_data = NULL;
 		goto install;
 	}
diff --git a/drivers/net/wireless/ath/ath12k/wmi.c b/drivers/net/wireless/ath/ath12k/wmi.c
index 52f0046..4179b55 100644
--- a/drivers/net/wireless/ath/ath12k/wmi.c
+++ b/drivers/net/wireless/ath/ath12k/wmi.c
@@ -3227,7 +3227,8 @@ int ath12k_wmi_vdev_install_key(struct ath12k *ar,
 
 	tlv = (struct wmi_tlv *)(skb->data + sizeof(*cmd));
 	tlv->header = ath12k_wmi_tlv_hdr(WMI_TAG_ARRAY_BYTE, key_len_aligned);
-	memcpy(tlv->value, arg->key_data, arg->key_len);
+	if (arg->key_data)
+		memcpy(tlv->value, arg->key_data, arg->key_len);
 
 	ath12k_dbg(ar->ab, ATH12K_DBG_WMI,
 		   "WMI vdev install key idx %d cipher %d len %d\n",
-- 
2.34.1

